test/final.test.js

```diff
@@ -0,0 +1,149 @@
+'use strict'
+const pino = require('..')
+const fs = require('fs')
+const { test } = require('tap')
+const { sleep } = require('./helper')
+
+test('replaces onTerminated option', async ({throws}) => {
+  throws(() => {
+    pino({
+      onTerminated: () => {}
+    })
+  }, Error('The onTerminated option has been removed, use pino.final instead'))
+})
+
+test('throws if not supplied a logger instance', async ({throws}) => {
+  throws(() => {
+    pino.final()
+  }, Error('expected a pino logger instance'))
+})
+
+test('throws if not supplied a handler', async ({throws}) => {
+  throws(() => {
+    pino.final(pino())
+  }, Error('expected a handler function'))
+})
+
+test('throws if not supplied logger with pino.destination or pino.extreme instance', async ({throws, doesNotThrow}) => {
+  throws(() => {
+    pino.final(pino(fs.createWriteStream('/dev/null')), () => {})
+  }, Error('only compatible with loggers with pino.destination and pino.extreme'))
+
+  doesNotThrow(() => {
+    pino.final(pino(pino.destination()), () => {})
+  })
+
+  doesNotThrow(() => {
+    pino.final(pino(pino.extreme()), () => {})
+  })
+})
+
+test('returns an exit listener function', async ({is}) => {
+  is(typeof pino.final(pino(pino.destination()), () => {}), 'function')
+})
+
+test('listener function immediately sync flushes when fired', async ({pass, fail}) => {
+  const dest = pino.destination('/dev/null')
+  var passed = false
+  dest.flushSync = () => {
+    passed = true
+    pass('flushSync called')
+  }
+  pino.final(pino(dest), () => {})()
+  await sleep(10)
+  if (passed === false) fail('flushSync not called')
```

Comment:
Can you verify that the destination is actually written? I trust SonicBoom tests up a certain point.

Reply:
it's verified here: 

https://github.com/pinojs/pino/pull/454/files#diff-21810649b385f6e84ec1653845cc5221R6

Reply:
the extreme mode test would fail without it